changed has_many through to use association when building query

johnpmitsch opened the pull request:
Fixes #131

domcleal commented:
The test passes on master as far as I can tell, it doesn't seem to demonstrate the bug.

It's hard to tell without a stack trace on the original bug report, but judging by the comment at https://github.com/wvanbergen/scoped_search/issues/131#issuecomment-109054530 and the `undefined method `klass' for nil:NilClass` error, I'd guess the actual bug lies on this line:

    middle_table_name = definition.reflection_by_name(many_class, through).klass.table_name

Though in your patch it's still:

    middle_table_class = definition.reflection_by_name(many_class, through).klass

Won't that still fail to look up the association, resulting in a `nil.klass` error?

johnpmitsch commented:
@domcleal 
That would make things easier if I included the full traceback :) here it is:

```2015-07-14T08:54:33 [app] [E] NoMethodError: undefined method `klass' for nil:NilClass
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:360:in `reflection_keys'
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:264:in `has_many_through_join'
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:235:in `sql_test'
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:500:in `sql_test'
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:457:in `to_single_field_sql'
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:469:in `to_sql'
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:51:in `build_find_params'
 | /home/vagrant/scoped_search/lib/scoped_search/query_builder.rb:22:in `build_query'
 | /home/vagrant/scoped_search/lib/scoped_search/definition.rb:255:in `block in register_named_scope!'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/scoping/named.rb:178:in `call'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/scoping/named.rb:178:in `block (2 levels) in scope'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/scoping/default.rb:41:in `block in unscoped'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/relation.rb:241:in `block in scoping'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/scoping.rb:98:in `with_scope'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/relation.rb:241:in `scoping'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/scoping/default.rb:41:in `unscoped'
 | /home/vagrant/.rvm/gems/ruby-1.9.3-p448/gems/activerecord-3.2.21/lib/active_record/scoping/named.rb:178:in `block in scope'
 | /home/vagrant/katello/app/controllers/katello/api/v2/api_controller.rb:81:in `scoped_search'
 | /home/vagrant/katello/app/controllers/katello/api/v2/content_view_versions_controller.rb:24:in `index'```

The '.klass' that is failing is actually in reflection keys https://github.com/wvanbergen/scoped_search/blob/master/lib/scoped_search/query_builder.rb#L360  which is being called from here https://github.com/wvanbergen/scoped_search/blob/master/lib/scoped_search/query_builder.rb#L264 

middle_table_name is passed with the module name attached and then returning nil for definition.reflection_by_name, which is passed to reflection keys as an argument.

It fails in the specific instance I found the bug because of namespaced modules which will be fixed by this PR.

domcleal commented:
Ah right, thanks, that makes more sense.

Do you know why the test is passing at the moment?  Is it a compatibility issue with AR 3.x only?  When I ran the test and debugged it earlier, the table name was always coming back without the module name.

johnpmitsch commented:
@domcleal which test do you mean?

johnpmitsch commented:
@domcleal oh you probably mean the test I wrote. I will look it out now and see

johnpmitsch commented:
@domcleal I think I see what is happening, I believe the module is in the table name because we are using an isolated engine (katello) and a module, so more specifically the error mentioned is returned while searching a has_many => through relationship on a namespaced model in an isolated engine. My test is not reflecting that, it just has the module. I am not entirely sure how I could reproduce an isolated engine in rspec, but looking into it now. Thanks for bringing that to my attention.

jlsherrill commented:
@johnpmitsch i think the reason your new test is passing against master are these lines:
```
+        module Zan; class Mar < ActiveRecord::Base; belongs_to :baz; belongs_to :koo; end; end
 +        module Zan; class Baz < ActiveRecord::Base; has_many :mars, :class_name => "Mar"; end; end 
```

The reason it is working is because 'mars' would resolve to 'Mar' normally.  You need to specify some other name, like ":other_mars". if you look at my modified example closely, that is what i did.

johnpmitsch commented:
@jlsherrill and I just discussed this off the thread and it seems there is two separate issues, but it should be possible to solve both. One is using a namespaced module in an engine which results in using a different table name, and the other is if the specified name is different than the table name (like :other_mars mentioned above) They both stem from this assuming the association name and the table name are the same.

johnpmitsch commented:
I updated this to have two tests to reflect the namespaced module issue and using a specified name in the relationship. Both are failing on master locally for me. I've changed my code to solve these two issues by checking for an association first in reflections, and using that or the middle_table_name if no association is found.

domcleal commented on lib/scoped_search/query_builder.rb:
```diff
@@ -246,6 +246,16 @@ def sql_test(field, operator, value, lhs, &block) # :yields: finder_option_type,
+    def find_has_many_through_association(field, through)
+      middle_table_association = nil
+      field.klass.reflect_on_all_associations(:has_many).each do |reflection| 
+        class_name = reflection.options[:class_name].downcase.pluralize if reflection.options[:class_name]
```
`constantize.table_name` might be more accurate than `downcase.pluralize` for CamelCase class names, which should be 'camel_cases' rather than 'camelcases' I think.

johnpmitsch replied:
just updated it thanks!

domcleal commented on spec/integration/relation_querying_spec.rb:
```diff
@@ -417,7 +417,113 @@ class Cat < ActiveRecord::Base
+        Moo.search_for('paz').length.should == 2
+      end
 
+      it "should find the two records that are related to a paz record" do
```
"find the one record that is related to two paz records"

domcleal commented on spec/integration/relation_querying_spec.rb:
```diff
@@ -417,7 +417,113 @@ class Cat < ActiveRecord::Base
+        Zan::Koo.search_for('baz').length.should == 2
+      end
+
+      it "should find the two records that are related to a baz record" do
```
ditto

johnpmitsch replied:
thanks, updated

domcleal commented:
Thanks @johnpmitsch!

wvanbergen commented:
Thanks @johnpmitsch & @domcleal
